### PR TITLE
Enable SolutionUserObject and SolutionFunction to evaluate gradients 

### LINKED ISSUE #7244: Enable evaluation of gradients in SolutionUserObject
### Description of the enhancement or error report

`SolutionUserObject` currently only evaluates point values but no gradients. `LibMesh::mesh_function` 
implements methods to retrieve the currents so it should be rather straight forward to enable 
`SolutionUserObject` to do the same.
### Rationale for the enhancement or information for reproducing the error

For neutron diffusion calculations, the gradient is very important for us because it allows
evaluation of the current and carries angular dependency information. 
### Identified impact

Current interfaces remain untouched. A new interface for accessing the gradient will be provided.

### PR DESCRIPTION
**Pull request by snschune**
Added interface for accessing the `SolutionUserObjects` gradient and a test. 

closes #7244

### CODE REVIEW

**Comment by snschune** [comment]
There is an issue if evaluated _on_ an element face, because gradients are not continuous when using 1st order shape functions. I assume you either get the left or right value depending on which element comes first in the  element_locater algorithm? How are DFEM shape functions dealt with in SolutionFunction?

Maybe I could add a check if it's close enough to an element boundary and return two values. Then define left and right depending on the element numbering? I am open for suggestions. Evaluation of 
element's faces is exactly the application I have in mind.

**Comment by YaqiWang** [comment]
If a side has a unique norm or orientation we can add a flag indicating evaluating plus or minus wrt the orientation .

**Comment by snschune** [comment]
Could even be a little worse - what if the point is indeed a mesh node that is shared by a number of elements?

File: framework/src/userobject/SolutionUserObject.C
Comment (permcody) [inline framework/src/userobject/SolutionUserObject.C]: You could just pass the Point in by value if you have to make a copy of it anyway.

File: framework/src/userobject/SolutionUserObject.C
Comment (permcody) [inline framework/src/userobject/SolutionUserObject.C]: spacing

File: framework/src/userobject/SolutionUserObject.C
Comment (permcody) [inline framework/src/userobject/SolutionUserObject.C]: spacing

File: framework/src/userobject/SolutionUserObject.C
Comment (permcody) [inline framework/src/userobject/SolutionUserObject.C]: spacing in each of the `i<LIBMESH_DIM` conditionals

File: framework/src/userobject/SolutionUserObject.C
Comment (permcody) [inline framework/src/userobject/SolutionUserObject.C]: spacing

**Comment by snschune** [comment]
@permcody This is not complete until I figure a way of how to handle discontinuous shape functions on faces. Can you set the `Do Not Merge` flag for now?

**Comment by YaqiWang** [comment]
While reviewing a paper, I found there is a theorem called Huilier theorem, which accept three directions and then evaluate the solid angle between three diametrical planes formed by the three directions. This solid angle can be used to weight the solutions on the elements surrounding points or lines.

**Comment by snschune** [comment]
This won't work without a libmesh update. I think the interface might need some fresh ideas as well. Currently, I just mimic the `mesh_function` interface here. These "raw" access APIs may be useful, but maybe we can think of simply weighting non-unique values or pick one by letting the user supply some rule (larger/smaller element id, distance of the centroid to some point, one element is in some solid type geometrical body or left/right of a sideset). There is some time to think about this till the next libmesh update so everyone who is interested can chime in.

**Comment by YaqiWang** [comment]
I

**Comment by snschune** [comment]
@YaqiWang The enumeration can be obtained from the number of 'hits' you get:

1 - interior
2 - element face
3+ - node

> While reviewing a paper, I found there is a theorem called Huilier theorem, which accept three directions and then evaluate the solid angle between three diametrical planes formed by the three directions. This solid angle can be used to weight the solutions on the elements surrounding points or lines.

Do you have a reference to this? When searching I find that  Huilier theorem is for computing the spherical excess of a spherical triangle. I am not quite sure how that helps us evaluate a suitable unique value on faces/nodes.

**Comment by snschune** [comment]
The last commit adds an example of how to wrap access to point values and gradients into a single function. The user has to provide a `weighting_type` that currently has options `average largest_element_id smallest_element_id`. Before adding any more complicated weightings I'd like to hear opinions on this implementation.

**Comment by snschune** [comment]
Can someone remove the `Do Not Merge` label. This is ready for review.

**Comment by jwpeterson** [comment]
Looks like this error message also needs to be updated as we aren't using `parallel_type` any more...

```
*** ERROR ***
Cannot use SolutionUserObject with DistributedMesh!
Consider specifying parallel_type = 'replicated' in your input file
to prevent it from being run with DistributedMesh.

Stack frames: 11
0: libMesh::print_trace(std::ostream&)
1: MooseMesh::errorIfDistributedMesh(std::string) const
2: SolutionUserObject::initialSetup()
3: MooseObjectWarehouse<GeneralUserObject>::initialSetup(unsigned int) const
4: FEProblem::initialSetup()
5: Steady::init()
6: MooseApp::executeExecutioner()
7: MooseApp::run()
8: /home/moosetest/civet/client_root_0/moose/test/moose_test-opt() [0x4064cd]
9: __libc_start_main
10: /home/moosetest/civet/client_root_0/moose/test/moose_test-opt() [0x406359]
[0] /home/moosetest/civet/client_root_0/moose/framework/src/mesh/MooseMesh.C, line 2313, compiled Oct 21 2016 at 10:34:17
```

**Comment by snschune** [comment]
@jwpeterson wait, I was going to change the error message, but I think it's correct. `parallel_type` is the new name, and `distribution` is deprecated.

**Comment by YaqiWang** [comment]
Can somebody review this? We will need to evaluate the solution gradient for our simulations. Thanks.

**Comment by permcody** [comment]
@andrsd - If you have a few minutes perhaps you can review this since you are the only one around! @YaqiWang, there is a conflict in this file. You'll need to rebase before we merge assuming there are no comments.

File: framework/include/functions/SolutionFunction.h
Comment (andrsd) [inline framework/include/functions/SolutionFunction.h]:
```
/**
 * Text ...
 */
```

File: framework/include/userobject/SolutionUserObject.h
Comment (andrsd) [inline framework/include/userobject/SolutionUserObject.h]: You are overloading virtual functions. That is not a good idea. Once somebody inherits from this, it will hide the other method in the child class. Either you do not expect people to override and then you do not need `virtual` or you do, but then pick unique names... Similarly below with other methods.

File: framework/include/userobject/SolutionUserObject.h
Comment (andrsd) [inline framework/include/userobject/SolutionUserObject.h]: Do you really want to return this by value? There will be a temporary copy, but I do know how big your maps will be...

File: test/include/auxkernels/FunctionGradAux.h
Comment (andrsd) [inline test/include/auxkernels/FunctionGradAux.h]: There is a real information value in this comment. Border-line with cpt. Obvious....

File: test/include/auxkernels/FunctionGradAux.h
Comment (andrsd) [inline test/include/auxkernels/FunctionGradAux.h]: This feels like it is a lie...

File: test/include/postprocessors/TestDiscontinuousValuePP.h
Comment (andrsd) [inline test/include/postprocessors/TestDiscontinuousValuePP.h]: I could not tell from the method signature this was a destructor...

File: test/include/postprocessors/TestDiscontinuousValuePP.h
Comment (andrsd) [inline test/include/postprocessors/TestDiscontinuousValuePP.h]: IMO it is not helpful to document c-tors of MooseObject-derived classes for the sake of having a doxy comment there... It is always the same comment over and over...

File: test/include/postprocessors/TestDiscontinuousValuePP.h
Comment (andrsd) [inline test/include/postprocessors/TestDiscontinuousValuePP.h]: The style for methods IIRC is:
```
/**
 *
 */
```

And again, doco strings for the sake of doco strings...

File: test/src/postprocessors/TestDiscontinuousValuePP.C
Comment (andrsd) [inline test/src/postprocessors/TestDiscontinuousValuePP.C]: `for_gradient` is a weird name... I know this is just a test class, but...

File: test/src/postprocessors/TestDiscontinuousValuePP.C
Comment (andrsd) [inline test/src/postprocessors/TestDiscontinuousValuePP.C]: `else` here would increase legibility...

File: test/tests/functions/solution_function/solution_function_grad_p2.i
Comment (andrsd) [inline test/tests/functions/solution_function/solution_function_grad_p2.i]: We need to stop copying this comment around! The same in the other test files...

File: test/tests/userobjects/solution_user_object/discontinuous_value_solution_uo_p1.i
Comment (andrsd) [inline test/tests/userobjects/solution_user_object/discontinuous_value_solution_uo_p1.i]: Move into `[ICs]` block

File: test/tests/userobjects/solution_user_object/discontinuous_value_solution_uo_p1.i
Comment (andrsd) [inline test/tests/userobjects/solution_user_object/discontinuous_value_solution_uo_p1.i]: And this as well...

File: test/src/auxkernels/FunctionGradAux.C
Comment (andrsd) [inline test/src/auxkernels/FunctionGradAux.C]: `MooseEnum` would be better. Would help Atom and peacock user. Conversion into an `unsigned int` would be trivial in the c-tor.

File: framework/src/functions/SolutionFunction.C
Comment (andrsd) [inline framework/src/functions/SolutionFunction.C]: spaces around `*`

**Review by andrsd — CHANGES_REQUESTED** [review]
And now that we have the moose doco system active, we do require a .md file for each new class, right?

File: framework/include/userobject/SolutionUserObject.h
Comment (snschune) [inline framework/include/userobject/SolutionUserObject.h]: This map is actually fairly small, because it only stores one value for each element that is adjacent to one point. This could be 8 if it's a corner in a hex mesh e.g.

File: test/tests/functions/solution_function/solution_function_grad_p2.i
Comment (snschune) [inline test/tests/functions/solution_function/solution_function_grad_p2.i]: oh ok, I thought it's the default anyway.

Comment (andrsd) [inline test/tests/functions/solution_function/solution_function_grad_p2.i]: Yes, it is. It is fine this way... You do not have to change it back ;-)

**Review by andrsd — APPROVED** [review]

**Comment by GiudGiud** [comment]
so this PR added a pointValueWrapper routine with some ideas for handling discontinuous routines.

I don't see any objects using it though? Anywhere in MOOSE Bison Pronghorn griffin now and then ?

**Comment by YaqiWang** [comment]
This is almost 10 years ago ;-) Based on the conversation, we would find the usage in Griffin. @GiudGiud I assume you added a mooseError in the function and did not see any test failure? Or test coverage tells us no-hit?

**Comment by GiudGiud** [comment]
coverage is down right but I searched pretty extensively

**Comment by YaqiWang** [comment]
Is there a way with git to search whether `pointValueGradientWrapper` existed in history elsewhere like in Griffin?

**Comment by GiudGiud** [comment]
there is git grep

**Pull request merged**
